Make clingo the default solver

alalazo opened the pull request:
Modifications:
- [x] Change `defaults/config.yaml`
- [x] Add a fix for bootstrapping patchelf from sources if `compilers.yaml` is empty
- [x] Make `SPACK_TEST_SOLVER=clingo` the default for unit-tests
- [x] Fix package failures in the e4s pipeline

Caveats:
1. CentOS 6 still uses the original concretizer as it can't connect to the buildcache due to issues with `ssl` (bootstrapping from sources requires a C++14 capable compiler)
1. I had to update the image tag for GitlabCI in e699f14. ~I did this since Radiuss is using that image and is already working with clingo, but it's not clear to me what are the differences between the two tags, so asked for @scottwittenburg review of that change.~ Thanks for the review @scottwittenburg 
1. libtool v2.4.2 has been deprecated and other packages received some update

alalazo committed:
Make clingo the default solver

- [x] Change defaults/config.yaml
- [x] Add a fix for bootstrapping patchelf from sources
      if compilers.yaml is empty
- [x] Make SPACK_TEST_SOLVER=clingo the default for unit-tests

alalazo committed:
Explicitly bootstrap before unit tests in centos6 and ubi

alalazo committed:
Revert CentOS 6 to use the original solver in CI

We cannot download information from the binary caches:

==> [2021-08-19-09:26:29.002062] Unable to read index hash https://mirror.spack.io/bootstrap/github-actions/v0.1/build_cache/index.json.hash
  Download failed: <urlopen error [Errno 1] _ssl.c:492: error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol>
  1

alalazo committed:
Update the default image, check if it makes a difference for clingo

scottwittenburg commented on share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml:
```diff
@@ -1,7 +1,7 @@
 
 default:
-  image: { "name": "ghcr.io/scottwittenburg/ecpe4s-ubuntu18.04-runner-x86_64:2020-09-01", "entrypoint": [""] }
+  image: { "name": "ghcr.io/scottwittenburg/ecpe4s-ubuntu18.04-runner-x86_64:2021-05-15", "entrypoint": [""] }
```
Did you already check that the compilers this image has are the same as the `2020-09-01` tagged image?  If not, pipeline jobs will end up doing a lot of duplicated bootstrapping.  So in that case, we would either need to change the compiler we specify in the stacks that use the default image (at least `share/spack/gitlab/cloud_pipelines/stacks/e4s/spack.yaml`, but also maybe others), or else make a new image with the desired compilers already installed.

alalazo replied:
No, I didn't. For build-systems it might not matter but I think it matters for e4s. I'll stop the pipeline.

alalazo replied:
@tgamblin Would it be fine with you if in this PR I add logic to use the original concretizer for the e4s pipeline, until #25287 is ready to be merged?

scottwittenburg replied:
Phew, that explains why it was canceled!  😅  I was worried there for a sec, when I didn't see any new push or request for spackbot to start a pipeline.

alalazo replied:
So, after a chat on slack with @eugeneswalker it seems to me that we'll be using the same compilers as before, so I'll restart the pipeline and see how many rebuilds it triggers.

alalazo commented:
@spackbot run pipeline

becker33 commented on etc/spack/defaults/config.yaml:
```diff
@@ -164,7 +164,7 @@ config:
   # 'clingo' currently requires the clingo ASP solver to be installed and
   # built with python bindings. 'original' is built in.
-  concretizer: original
+  concretizer: clingo
```
We should update the comment here.

alalazo replied:
Comment updated in 549f446 Let me know if that's good or should be rephrased.

becker33 commented on lib/spack/spack/bootstrap.py:
```diff
@@ -418,6 +418,14 @@ def _bootstrap_config_scopes():
 
 @contextlib.contextmanager
 def ensure_bootstrap_configuration():
+    # We may need to compile code from sources, so ensure we have compilers
+    # for the current platform before switching parts. We use "spack compiler
+    # list" as it searches for compilers lazily i.e. only if we don't have
+    # them already in compilers.yaml
+    compiler_cmd = spack.main.SpackCommand('compiler')
+    compiler_cmd(
+        'list', output=os.devnull, error=os.devnull, fail_on_error=False
+    )
```
This doesn't actually ensure we have any compilers available for the current architecture -- if you have only compilers for a different OS, `spack compiler list` still won't find new ones.

We should explicitly call into the `spack.compilers` module and check whether we have any compilers for the current architecture.

alalazo replied:
Updated the logic in 96ac691. Spack looks into `compilers.yaml` first and calls:
```console
spack compiler find
```
only if we don't have a compiler for the current host.

alalazo committed:
config.yaml: update the 'concretizer' comment

alalazo committed:
Strengthen a bit initialization of compilers

becker33 commented on lib/spack/spack/bootstrap.py:
```diff
@@ -416,8 +416,30 @@ def _bootstrap_config_scopes():
     return config_scopes
 
 
+def _has_compiler_for_host():
+    """True if the current configuration has at least one compiler for the host"""
+    host = spack.architecture.default_arch()
+    host_target, host_os = str(host.target.microarchitecture.family), str(host.os)
+    compilers = spack.compilers.get_compiler_config(init_config=False)
+    for entry in compilers:
+        compiler = entry['compiler']
+        compiler_is_for_host = (compiler['operating_system'] == host_os and
+                                compiler['target'] == host_target)
+        if compiler_is_for_host:
+            return True
+    return False
+
+
 @contextlib.contextmanager
 def ensure_bootstrap_configuration():
+    # We may need to compile code from sources, so ensure we have compilers
+    # for the current platform before switching parts.
+    if not _has_compiler_for_host():
```
```suggestion
@contextlib.contextmanager
def ensure_bootstrap_configuration():
    # We may need to compile code from sources, so ensure we have compilers
    # for the current platform before switching parts.
    arch = spack.architecture.default_arch()
    if not spack.compilers.compilers_for_arch(arch):
```

alalazo replied:
Done, thanks!

alalazo committed:
Simplify code to find compilers if none is there

tldahlgren commented on lib/spack/spack/environment.py:
```diff
@@ -1703,7 +1703,22 @@ def matching_spec(self, spec):
         # Dependency-only specs will have value None
         matches = {}
 
+        if not isinstance(spec, spack.spec.Spec):
+            spec = spack.spec.Spec(spec)
+
         for user_spec, concretized_user_spec in self.concretized_specs():
+            # Deal with concrete specs differently
+            if spec.concrete:
+                # Matching a concrete spec is more restrictive
+                # than just matching the dag hash
+                is_match = (
+                    spec in concretized_user_spec and
+                    concretized_user_spec[spec.name].build_hash() == spec.build_hash()
+                )
+                if is_match:
+                    matches[spec] = spec
+                continue
+
```
Can this fix be moved to a different PR since it applies for Cloud CI whether `clingo` is the default or explicitly requested by the user?

I'd also hate to see this fix delayed since I've run into the problem for at least one, maybe two different CI efforts.

alalazo replied:
Done and merged! Thanks!

alalazo committed:
Fix e4s stack build with clingo

Remove preferences that may direct the solver
towards libtool@develop

Add a conflict to numactl

alalazo committed:
libtool: deprecate version 2.4.2

alalazo commented:
@becker33 @tgamblin Tests are passing, let me know if there are further modifications to be done here.

haampie commented on lib/spack/spack/bootstrap.py:
```diff
@@ -418,6 +418,16 @@ def _bootstrap_config_scopes():
+    arch = spack.architecture.default_arch()
+    arch = spack.spec.ArchSpec(str(arch))  # The call below expects an ArchSpec object
+    if not spack.compilers.compilers_for_arch(arch):
+        compiler_cmd = spack.main.SpackCommand('compiler')
```
are we using SpackCommand in our code directly now?

> Callable object that invokes a spack command (for testing).

is in its docstring

tgamblin replied:
Would rather not use it in the code, but in a pinch we can.  It would be better to factor this so that it does not go through `SpackCommand` IMO.

alalazo replied:
We use `SpackCommand` in two places in `bootstrap.py`. The fact is our cli commands are not always thin wrappers of API calls but they may have a fair amount of logic in them. I agree this should be refactored, but I'd prefer to do it in a PR focused on refactoring the API - so that it wouldn't be out of place refactoring other related calls in an internal module and the PR would be focused on a single objective.

alalazo commented:
@spackbot run pipeline

scottwittenburg commented:
fyi @eugeneswalker

tldahlgren reviewed (dismissed):
LGTM.

becker33 commented on .github/workflows/unit_tests.yaml:
```diff
@@ -225,23 +225,25 @@ jobs:
           git fetch origin ${{ github.ref }}:test-branch
           git checkout test-branch
-          share/spack/qa/run-unit-tests
+          bin/spack unit-test -x -m "not maybeslow" -k "package_sanity"
```
Is there a reason we're not running the shell tests anymore?

alalazo replied:
The libtool change was a quick workaround a month ago to not select libtool 2.4.2 in DAGs.  The real issue has been then fixed in https://github.com/spack/spack/pull/25585 but since the package update also made sense in itself I kept it. I can extract that into another PR if it makes more sense.

alalazo replied:
> Is there a reason we're not running the shell tests anymore?

It's because of point 1. in the description (note that we are not running shell tests anymore only in the CentOS 6 tests):

> CentOS 6 still uses the original concretizer as it can't connect to the buildcache due to issues with ssl (bootstrapping from sources requires a C++14 capable compiler)

I encountered issues with CentOS 6 in CI due apparently to using an old Python interpreter. Since it is my understanding (or maybe it's just a strong hope :grimacing: ) that the second after we release v0.17.0 we also drop Python 2.6 I worked around that issue by using the original concretizer and avoiding the:
```console
$ spack spec mpileaks
```
we have in the fast shell tests. Honestly I didn't investigate further the SSL error that was reported with Python 2.6, but let me know if I should dig deeper into it.

becker33 reviewed (commented):
Are the libtool/numactl changes really a part of this PR? Is that something that currently fails in clingo?

Mostly LGTM.

alalazo commented:
@spackbot run pipeline

alalazo commented:
@spackbot run pipeline

alalazo commented:
@spackbot run pipeline
